### PR TITLE
feat(orders): BACK-617 add backorder display to re-order dialog

### PR DESCRIPTION
**Pull request by declankirk**
Jira: [BACK-617](https://bigcommercecloud.atlassian.net/browse/BACK-617)

## What/Why?
When backorder features are enabled, the order re-order dialog loads variant inventory from `getVariantInfoBySkus` (including `totalOnHand` and `backorderMessage`) and shows the same backorder lines as quotes do (i.e. ATS, backorder amount, backorder message).

## Rollout/Rollback
Merge/revert. Backorder display is gated by `Feature_Backorder && BACK-134.backorders_phase_1_1_control_messaging_on_storefront`.

## Testing
Backorder display appears as expected for a variety of cases, updates with quantity.

https://github.com/user-attachments/assets/2b41ef65-8eb9-4402-ac02-3a88c6428783

Order with no backordered quantity does not display backorder lines.

Screenshot 2026-05-13 at 10 50 30 am

Lines do not appear when `Feature_Backorder && BACK-134.backorders_phase_1_1_control_messaging_on_storefront` disabled.

Screenshot 2026-05-13 at 10 58 45 am

Regression - backorder display still appears as expected for both `quoteDraft` and `quoteDisplay`.

Screenshot 2026-05-14 at 11 03 52 am

Screenshot 2026-05-14 at 11 04 17 am

Ping @animesh1987 @benpratt77

[BACK-617]: https://bigcommercecloud.atlassian.net/browse/BACK-617?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---

> [!NOTE]
> **Medium Risk**
> Adds inventory/backorder fields to the `getVariantInfoBySkus` GraphQL query and uses them to conditionally render backorder messaging in the re-order flow; incorrect query/typing or gating logic could impact ordering UI and data shown to customers.
> 
> **Overview**
> Adds a new `useBackorderStorefrontMessaging` hook (with tests) to centralize storefront backorder messaging gating via `Feature_Backorder`, the `BACK-134...` flag, and backorder display settings.
> 
> Extends the order re-order dialog to fetch per-SKU inventory (`totalOnHand`, `backorderMessage`, etc.) via `getVariantInfoBySkus`, compute backorder display fields, and render `BackorderMessage` beneath the quantity input when enabled.
> 
> Refactors quote backorder gating to use the new hook, and unifies backorder field computation by introducing `getBackorderDisplayFieldsFromOnHand` and reusing it for draft quotes and re-order display (with added unit tests).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 342e0577ffc5c0a8be6d4ac2d454b5c00b74a7f9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx [hunk 1]
```diff
@@ -184,71 +196,113 @@ export default function OrderCheckboxProduct(props: OrderCheckboxProductProps) {
...
-              }}
-              error={!!product.helperText}
-              helperText={product.helperText}
+      {products.map((product: EditableProductItem) => {
+        const qty = Number(getProductQuantity(product)) || 0;
+        const invRow = reorderInventoryBySku?.[product.sku.toUpperCase()];
+        const backorderFields = getReorderBackorderDisplayFields(qty, invRow);
+        const showReorderBackorder =
+          reorderBackorderUiEnabled && type === 'reOrder' && Boolean(backorderFields);
+
```
[the remark this answers is not in the document]
Comment (declankirk) [reply apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx:204]: This is something that will be addressed when we implement quantity error handling in [BACK-634](https://bigcommercecloud.atlassian.net/browse/BACK-634)

File: apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx [hunk 2]
```diff
@@ -59,6 +66,11 @@ export default function OrderCheckboxProduct(props: OrderCheckboxProductProps) {
   };
 
   const itemStyle = isMobile ? mobileItemStyle : defaultItemStyle;
+  const qtyStackAlignItems = textAlign === 'right' ? 'flex-end' : 'flex-start';
+  const desktopQtyColumnStyle =
+    reorderBackorderUiEnabled && type === 'reOrder' && !isMobile
```
Comment (bc-gary) [inline apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx]: [Suggestion] **`type === 'reOrder'` check is duplicated across parent and child**

`OrderDialog` already gates `reorderBackorderUiEnabled` on `type === 'reOrder'` (line 543), but this component re-checks `type === 'reOrder'` in both `desktopQtyColumnStyle` and `showReorderBackorder`. Since the parent already guarantees the condition, the child could trust the prop without rechecking. Not a bug — just redundant guards.

File: apps/storefront/src/pages/OrderDetail/utils/reorderBackorderDisplay.test.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  it('getReorderBackorderDisplayFields returns null when no backordered quantity', () => {
+    const row = {
+      inventoryTracking: 'variant',
+      totalOnHand: 10,
+    } as CatalogQuickVariantSku;
+
+    expect(getReorderBackorderDisplayFields(10, row)).toBeNull();
+  });
+});
```
Comment (bc-gary) [inline apps/storefront/src/pages/OrderDetail/utils/reorderBackorderDisplay.test.ts:76]: [Suggestion] **Consider adding edge case coverage**

The two tests cover the happy path (qty > onHand) and the no-backorder case (qty === onHand). A few real-world edge cases would strengthen coverage:

- `inventoryTracking: 'none'` → should return null (products without inventory tracking)
- `row` is `undefined` → should return null
- `totalOnHand` is `null` → should return null
- `quantity` is 0 → should return null
- negative `totalOnHand` → entire quantity is backordered

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx [hunk 4]
```diff
@@ -461,6 +466,16 @@ export default function OrderDialog({
...
   };
 
+  const reorderInventoryBySku = useMemo(() => {
+    const map: Record<string, CatalogQuickVariantSku> = {};
+    variantInfoList.forEach((row) => {
+      if (row.variantSku) {
+        map[row.variantSku.toUpperCase()] = row as CatalogQuickVariantSku;
+      }
+    });
+    return map;
```
Comment (bc-gary) [inline apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx:473]: [Question] **`variantInfoList` typed as `CustomFieldItems[]` but cast to `CatalogQuickVariantSku`**

`variantInfoList` is `CustomFieldItems[]` (`Record<string, any>`) and each row is cast here with `as CatalogQuickVariantSku`. The inventory fields (`totalOnHand`, `inventoryTracking`, etc.) are trusted to exist because they're in the query, but the types don't enforce it. Is there a plan to type `getVariantInfoBySkus` return more precisely, or is the `CustomFieldItems` pattern intentional across the codebase?

Comment (declankirk) [reply apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx:473]: Will tighten the state here to `useState<CatalogQuickVariantSku[]>([])`

File: apps/storefront/src/pages/OrderDetail/utils/reorderBackorderDisplay.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  row: CatalogQuickVariantSku | undefined,
+): ReorderBackorderDisplayFields | null {
+  if (!row) return null;
+  const tracking = row.inventoryTracking || 'none';
+  if (tracking !== 'product' && tracking !== 'variant') return null;
+
+  const { totalOnHand } = row;
+  if (totalOnHand === null || totalOnHand === undefined) return null;
+
+  const quantityBackordered = Math.max(0, quantity - totalOnHand);
```
[the remark this answers is not in the document]
Comment (declankirk) [reply apps/storefront/src/pages/OrderDetail/utils/reorderBackorderDisplay.ts]: As above, this is something that will be addressed when we implement quantity error handling in [BACK-634](https://bigcommercecloud.atlassian.net/browse/BACK-634)

File: apps/storefront/src/shared/service/b2b/graphql/product.ts [hunk 6]
```diff
@@ -11,6 +11,14 @@ interface ProductPurchasable {
   sku: string;
 }
 
+export type CatalogQuickVariantSku = CustomFieldItems & {
+  inventoryTracking?: string;
+  availableToSell?: number;
+  unlimitedBackorder?: boolean;
+  totalOnHand?: number | null;
+  backorderMessage?: string | null;
```
Comment (bc-gary) [inline apps/storefront/src/shared/service/b2b/graphql/product.ts:20]: [Suggestion] **`variantSku` not declared in `CatalogQuickVariantSku`**

`OrderDialog`'s `reorderInventoryBySku` memo accesses `row.variantSku` to build the lookup map, and the GraphQL query requests `variantSku`, but the type doesn't declare it — it only compiles because `CustomFieldItems` (`Record<string, any>`) allows anything. If the `CustomFieldItems` intersection is ever removed to fully type this, the code would break silently.

Consider adding `variantSku?: string` to the type alongside the other inventory fields.

**Review by bc-gary — APPROVED** [review]

File: apps/storefront/src/pages/OrderDetail/utils/reorderBackorderDisplay.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import type { CatalogQuickVariantSku } from '@/shared/service/b2b/graphql/product';
+
+interface ReorderBackorderDisplayFields {
+  totalOnHand: number;
+  quantityBackordered: number;
+  backorderMessage?: string;
+}
+
+export function getReorderBackorderDisplayFields(
```
Comment (bc-gauravsobti) [inline apps/storefront/src/pages/OrderDetail/utils/reorderBackorderDisplay.ts:5]: [Question]
Since the logic is same, should we reuse `getDraftBackorderDisplayFields` here? By maybe creating a core helper for both?

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx [hunk 8]
```diff
@@ -524,6 +539,10 @@ export default function OrderDialog({
             setReturnArr={setReturnArr}
             textAlign={isMobile ? 'left' : 'right'}
             type={type}
+            reorderInventoryBySku={reorderInventoryBySku}
+            reorderBackorderUiEnabled={
+              isBackorderEnabled && isBackorderMessagingEnabled && type === 'reOrder'
```
Comment (bc-gauravsobti) [inline apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx]: [Required]
`isBackorderEnabled && isBackorderMessagingEnabled` is being used at a quite a few places now, it should be extracted into a custom hook so that it is to maintain and cleanup later.

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx [hunk 9]
```diff
@@ -524,6 +539,10 @@ export default function OrderDialog({
...
+            }
```
Comment (bc-gauravsobti) [inline apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx:555]: [Question]
`const hasAnyBackorderDisplay =
  showQuantityOnBackorder || showQuantityOnHand || showBackorderMessage;`

This was used in QuoteTable. Is there a need to use this here?

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx (hunk 8 above)
Comment (bc-gauravsobti) [reply apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx]: I think all the helpers related to backorder like this could be part of one custom hook.

**Review by bc-gauravsobti — CHANGES_REQUESTED** [review]

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx (hunk 8 above)
Comment (declankirk) [reply apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx]: Excellent suggestion, we should have done that ages ago 😮‍💨 

Will add a `useBackorderStorefrontMessaging` hook, and some smaller utils for computing the backorder display fields.

File: apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx [hunk 10]
```diff
@@ -43,6 +48,8 @@ export default function OrderCheckboxProduct(props: OrderCheckboxProductProps) {
     setReturnArr = () => {},
     textAlign = 'left',
     type,
+    reorderInventoryBySku,
+    reorderBackorderUiEnabled = false,
```
Comment (bc-gauravsobti) [inline apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx:52]: [Question]
Should this be a part of props or should we use the hook directly inside the function?

Comment (declankirk) [reply apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx:52]: IMO it makes sense to keep it as a prop here, so the parent `OrderDialog` composes the dialog mode plus `useBackorderStorefrontMessaging()`, and the list stays a simple presentational component without hidden feature flag dependencies.

Comment (bc-gauravsobti) [reply apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx:52]: This suggestion is not that important. Just something good to discuss for our understanding. 
I feel since we extracted `useBackorderStorefrontMessaging()`, `OrderCheckoutProduct` will still be a simple presentation component. Doing it this way just simplifies dependence with the parent by 1 prop which is not much of an improvement but we can also notice that we currently are passing 10 props to the child.

**Review by bc-gauravsobti — APPROVED** [review]
LGTM 🥳 
Amazing work.

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx [hunk 11]
```diff
@@ -461,6 +463,16 @@ export default function OrderDialog({
     setOpenShoppingList(true);
   };
 
+  const reorderInventoryBySku = useMemo(() => {
+    const map: Record<string, CatalogQuickVariantSku> = {};
+    variantInfoList.forEach((row) => {
```
Comment (bc-jeffteng) [inline apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx:468]: [suggestion]
`variantInfoList` is never cleared when the dialog closes or the order/products change, so `reorderInventoryBySku` can briefly reuse the previous inventory map on the next open. Can we reset it on close/open before fetching, or key the dialog by `orderId` (force recreate dialog component)?

**Review by benpratt77 — APPROVED** [review]

**Review by bc-jeffteng — APPROVED** [review]

**Comment by bc-launchbay** [comment]
*b2b-buyer-portal* deployed to Staging GCS Upload
success [20260517233104](https://launchbay.bigcommerce.net/projects/515/releases/285102) by Launchbay

**Comment by bc-launchbay** [comment]
*b2b-buyer-portal* deployed to Production GCS Upload
success [20260517233104](https://launchbay.bigcommerce.net/projects/515/releases/285102) by Launchbay
